tao_compiler/mlir/disc/transforms/lhlo_legalize_roots_to_loops.cc

```diff
@@ -1922,36 +2000,545 @@ LogicalResult lowerWithScheduleColReductionBlockTileSchedule(
+      auto updated = op;
+      // Alloc for, if necessary, and update output operands.
+      if (regular_xroots.contains(op) ||
+          (irregular_xroots.contains(op) &&
+           !visited_irregular_xroots.contains(op))) {
+        // Update input operands.
+        // Replace input with new allocated ones in previous iterations.
+        for (int64_t j = 0; j < num_input_operand; j++) {
+          auto operand = updated->getOperand(j);
+          if (auto new_operand = cloning_map.lookup(operand)) {
+            updated->replaceUsesOfWith(operand, new_operand);
+          }
+        }
+        // For non-sub-root xroot op, it should be write back during lowering.
+        if (!sub_roots.contains(op)) {
+          lower_config.setWrittenBack(op);
+        }
+      } else {
+        // For to-be-cloned ops, alloc for output and then clone the op with new
+        // outputs.
+        SmallVector<Value> results;
+        for (Value v : op->getOperands().drop_front(num_input_operand)) {
+          Value new_operand = allocClonedValue(v);
+          cloning_map.map(v, new_operand);
+          results.push_back(v);
+        }
+        updated = builder.clone(*op, cloning_map);
+        for (auto result : results) {
+          fusion_pattern.updateLastWriter(result, updated);
+        }
+        if (!irregular_xroots.contains(op)) {
+          to_erase.insert(op);
+        }
+      }
+      // Make sure the ops are in order.
+      if (last_op != nullptr) {
+        updated->moveAfter(last_op);
+      }
+      last_op = updated;
+    }
+    visited_irregular_xroots.insert(
+        skeleton_group.irregular_root_member_set.begin(),
+        skeleton_group.irregular_root_member_set.end());
+  }
+  // Clean been cloned ops. Note xroot ops are not to be erased.
+  for (auto op : to_erase) {
+    op->erase();
+  }
+
+  return success();
+}
+
+LogicalResult lowerWithScheduleStitch(lmhlo::FusionOp& fusion_op,
+                                      FusionPattern& fusion_pattern,
+                                      ShapeAnalysis* shape_analysis,
+                                      int64_t row_tile,
+                                      LowerConfig& lower_config) {
+  auto root_ops = fusion_pattern.getRootOps();
+  auto sub_root_ops = fusion_pattern.getSubRootOps();
+  auto result_values = fusion_pattern.getResults();
+  DenseSet<Operation*> roots(root_ops.begin(), root_ops.end());
+  // DenseSet<Operation*> sub_roots(sub_root_ops.begin(), sub_root_ops.end());
+  DenseSet<Value> results(result_values.begin(), result_values.end());
+  auto op_list = fusion_pattern.getOpList();
+  DenseSet<Operation*> op_set(op_list.begin(), op_list.end());
+  SmallVector<Operation*> external_only_root_ops;
+  for (auto value : fusion_pattern.getExternalOnlyResults()) {
+    external_only_root_ops.push_back(fusion_pattern.findLastWriter(value));
+  }
+  DenseSet<Operation*> external_only_roots(external_only_root_ops.begin(),
```

Comment:
why not directly operate on DenseSet?   
`DenseSet<Operation*> external_only_roots;
  for (auto value : fusion_pattern.getExternalOnlyResults()) {
    external_only_roots.push_back(fusion_pattern.findLastWriter(value));
  }`
Is this same as well?

Reply:
Fixed, thanks. I did not notice that `external_only_root_ops` is not useful any more...